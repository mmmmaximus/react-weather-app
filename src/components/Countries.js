import React from "react"
import Country from './Country'

const Countries = ({countries, filter, showButton}) => {
  const filteredCountries = countries.filter(
    country => country.name.toLowerCase().includes(filter.toLowerCase())
  )
  
  if (filter === '') {
    return 'Search for a country'
  } else if (filteredCountries.length > 10) {
    return 'Too many matches, specify another filter'
  } else if (filteredCountries.length <= 10 && filteredCountries.length > 1) {
    return <div>
      {filteredCountries.map(country => 
        <div>
          <p key={country.name}>{country.name} <button onClick={showButton} value={country.name}>show</button></p>
        </div>
      )}
    </div>
  } else {
    return <div>
      {filteredCountries.map(country => 
        <Country country={country}/>
      )}
    </div>
  }
}

export default Countries
