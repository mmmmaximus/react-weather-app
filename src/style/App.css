* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

html {
  height: 100%;
  font-family: "Lucida Sans", sans-serif;
}

body {
  background-image: linear-gradient(to top, #2362bf, #007cd1, #0095de, #00ade8, #06c4ee);
  color: #ffffff;
}

.filter {
  padding: 20px 50px;
  text-align: center;
}

.countries {
  padding: 20px 50px;
}

.countries > .filter-text {
  text-align: center;
}

.countries > .countries-text {
  margin-left: 38%;
}

.country {
  display: flex;
  justify-content: space-between;
}

.country > div {
  width: 48%;
  padding: 10px;
}

.country-data > .img {
  width: 100%;
  max-width: 350px;
  margin-left: auto;
  margin-right: auto;
  display: block;
  padding: 20px;
}

.country-data > h2 {
  text-align: center;
  font-size: 1.5em;
  padding: 5px 0px 5px;
}

.country-data > table {
  margin: auto;
}

tr > td {
  padding: 5px 2px;
}

.languages > h3 {
  padding: 10px 0px;
  text-align: center;
}

div > li {
  padding: 0px 25px;
  margin-left: 35%;
}

.weather > .img {
  width: 85%;
  max-width: 300px;
  margin-left: auto;
  margin-right: auto;
  display: block;
  padding: 0px 20px;
}

.weather > h2 {
  text-align: center;
  font-size: 1.5em;
  padding: 0px 0px 5px 0px;
}

h3.description {
  text-align: center;
  padding: 5px 0px;
}

.weather > table {
  margin: auto;
}
