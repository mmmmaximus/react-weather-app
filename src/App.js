import React, {useState, useEffect} from 'react';
import axios from 'axios';
import Filter from './components/Filter'
import Countries from './components/Countries'
import './style/App.css'

const App = () => {
  const [filter, setFilter] = useState('')
  const [countries, setCountries] = useState([])

  const changeFilter = (event) => {
    setFilter(event.target.value)
  }

  const showButton = (event) => {
    setFilter(event.target.value)
  }

  useEffect(() => {
    axios
      .get('https://restcountries.eu/rest/v2/all')
      .then(response => {
        setCountries(response.data)
      })
  }, [])

  return (
    <div data-testid='App'>
      <Filter func={changeFilter}/>
      <Countries countries={countries} filter={filter} showButton={showButton}/>
    </div>
  )
}

export default App;
