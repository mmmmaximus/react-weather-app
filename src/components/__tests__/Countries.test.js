import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import Countries from '../Countries'
import App from '../../App'
import renderer from 'react-test-renderer'
import mockedAxios from '../../__mocks__/axios'

const countriesResponse = {
  data: [
    {
      name: 'Africa',
      capital: 'Wakanda',
      population: '123456',
      languages: [
        {name: 'language1'},
        {name: 'language2'},
        {name: 'language3'}
      ]
    },
    { name: 'Afric' },
    { name: 'Afri' },
    { name: 'Afr' },
    { name: 'Af' },
    { name: 'A' },
    { name: 'frica' },
    { name: 'rica' },
    { name: 'ica' },
    { name: 'ca' },
    { name: 'fria' }
  ]
}
let filter = ''
const showButton = (event) => {
  filter = event.target.value
}
const countries = []

test('should render component', () => {
  render(<Countries countries={countries} filter={filter} showButton={showButton} />)
  const CountriesElement = screen.getByTestId('Countries')
  expect(CountriesElement).toBeInTheDocument
})

test('should filter countries when there is no filter applied', () => {
  mockedAxios.get.mockResolvedValueOnce(countriesResponse)
  render(<App />)
  const CountriesElement = screen.getByTestId('Countries')
  const FilterElement = screen.getByTestId('Filter')

  fireEvent.change(FilterElement, { target: { value: '' } })
  expect(CountriesElement.textContent).toBe('Search for a country')
})

test('should filter countries when filter applied and > 10 matches', async () => {
  mockedAxios.get.mockResolvedValueOnce(countriesResponse)
  render(<App />)
  const CountriesElement = screen.getByTestId('Countries')
  const FilterElement = screen.getByTestId('Filter')

  fireEvent.change(FilterElement, { target: { value: 'a' } })
  await waitFor(() => {
    expect(CountriesElement.textContent).toBe('Too many matches, specify another filter')
  })
})

test('should filter countries when filter applied and <= 10 but > 1 matches', async () => {
  mockedAxios.get.mockResolvedValueOnce(countriesResponse)
  render(<App />)
  const CountriesElement = screen.getByTestId('Countries')
  const FilterElement = screen.getByTestId('Filter')

  fireEvent.change(FilterElement, { target: { value: 'afri' } })
  await waitFor(() => {
    expect(CountriesElement.textContent).toContain('Africa')
    expect(CountriesElement.textContent).toContain('Afric')
    expect(CountriesElement.textContent).toContain('Afri')
  })
})

test('should filter countries when filter applied and 1 match', async () => {
  mockedAxios.get.mockResolvedValueOnce(countriesResponse)
  render(<App />)
  const FilterElement = screen.getByTestId('Filter')

  const weatherResponse = {
    data: {
      data: [null]
    }
  }
  mockedAxios.get.mockResolvedValueOnce(weatherResponse)
  fireEvent.change(FilterElement, { target: { value: 'africa' } })
  await waitFor(() => {
    const CountryElement = screen.getByTestId('Country')
    expect(CountryElement).toBeInTheDocument
  })
})

test('matches snapshot', () => {
  const tree = renderer.create(<Countries countries={countries} filter={filter} showButton={showButton} />).toJSON()
  expect(tree).toMatchSnapshot()
})
