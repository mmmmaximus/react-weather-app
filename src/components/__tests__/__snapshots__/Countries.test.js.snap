// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="countries"
  data-testid="Countries"
>
  Search for a country
</div>
`;
